### ISSUE TITLE
dynamic-require obscures an object's class

### ISSUE THREAD

**Issue by klmanion** [racket/racket#2216]
This error, present in Racket v6.10, was fixed in v6.12, but it reappears in v7.0

The following code, run in v7.0, produces the output provided in the comments, and the error message below it.
```
(provide (contract-out
           [read-map (module-path? . -> . ship?)]))

(define read-map
  (λ (m)
    (let ([ship (dynamic-require m 'ship)])
      (send ship place-rooms)
      (printf "~a\n" ship) ; #(struct:object:ship% ...)
      (printf "~a\n" (ship? ship)) ; #f
      ship)))
```
```
; read-map: broke its own contract
;   promised: ship?
;   produced: (object:ship% ...)
;   in: the range of
;       (-> module-path? ship?)
```
On v6.12 the program compiles error free with the ship object being correctly apprehended as such.
(`ship?` is nothing more than a wrapper over a `is-a?` predicate passed the ship interface in 1st position and `ship?`'s argument in 2nd.)

**Comment by mflatt** [comment]
Hi @klmanion  – can you provide a full example to run? There's not enough in this fragment for me to guess at what the problem may be, since I don't know enough about the namespace setup or what form of module reference `m` may take.

**Comment by klmanion** [comment]
@mflatt of course.  `git clone -b draw_room https://www.github.com/klmanion/within.git` should do the trick.

The fragment comes from the file src/langs/map/read-map.rkt.  The module `m` refers to a module declared with the lang line `#lang map` such as maps/test1.rkt, which is the file passed to `read-map` by the canvas ([src/game-canvas.rkt:26](https://github.com/klmanion/within/blob/7394fbc8af818bba467674063f99ddffd8a3bb5e/src/game-canvas.rkt#L26)) as a string.  The implementation of this canvas is in src/main.rkt.

The map lang describes a map for this game, saving it in a ship object.

The current commit, klmanion/within@7394fbc, builds correctly in v6.12.


Thanks for the timely reply!

**Comment by mflatt** [comment]
I'm still unclear on your setup, but my guess is that you have the `map` collection mapped to a different instance of the code than the one in "src/langs/map" adjacent to the "src/main.rkt" that you're running.

I tried installing "src/langs/map" as a package, so that `#lang map` refers to the language in that directory. Then, when I change `src/lang/maps/read-map` to

```
#lang racket/base
(require racket/class
         "../../ship.rkt")

(provide read-map)

(define read-map
  (λ (m)
    (define v (dynamic-require m 'ship))
    (log-error "~s" (is-a? v ship<%>))
    v))
```

and run "src/main.rkt", the logged output is `#t`. But if I clone the repo (with that change) to a fresh location and run "src/main.rkt" there, then the logged output is `#f`, because `#lang map` refers to a different "ship.rkt" than "src/main.rkt" in the clone.

It makes sense that you'd see a problem with v7.0 and not v6.12, because package and collection registrations are version-specific. The directory where you're running must be mapped to the `map` collection for your v6.12 installation, with some different directory mapped to `map` for your v7.0 installation.

**Comment by klmanion** [comment]
You seem to be right: I switched to v7.0, removed the package, installed it again, and it worked like a charm.

I'm still a bit unclear though as to why this works.  What ship.rkt is the old installation of the `map` package referencing?  And is this something I should be concerned about when writing a makefile in the future?
